### PR TITLE
terraform: Terraform module for GCP

### PR DESCRIPTION
**Pull request by elchead**
### Context

This PR builds on top of #2503 to add the module for GCP.

### Proposed change(s)

- add GCP module

### Additional info

- [AB#3520](https://dev.azure.com/Edgeless/ae37573d-ccde-4af2-ab1e-001e587197d1/_workitems/edit/3520)
- [successful GCP E2E run](https://github.com/edgelesssys/constellation/actions/runs/6782739265/job/18435569332)
- [AWS E2E run](https://github.com/edgelesssys/constellation/actions/runs/6783103693)

### Checklist

- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [x] Link to Milestone

### CODE REVIEW

**Comment by msanft** [comment]
Let's put this on hold until #2503 is merged.

File: terraform/constellation-cluster/main.tf
Comment (msanft) [inline terraform/constellation-cluster/main.tf:63]:
```suggestion
```

File: terraform/gcp-constellation/main.tf
Comment (msanft) [inline terraform/gcp-constellation/main.tf:46]:
```suggestion
resource "null_resource" "service_account_file" {
```
sa_account is semantically duplicate (i.e. serviceaccount_account)

File: terraform/gcp-constellation/main.tf
Comment (msanft) [inline terraform/gcp-constellation/main.tf:49]: what exactly does this mean?

**Review by msanft — COMMENTED** [review]
LGTM from a high level. Will give this a more thorough review once the AWS module is merged.

File: terraform/gcp-constellation/main.tf
Comment (elchead) [reply terraform/gcp-constellation/main.tf:49]: I originally wanted to try to get rid of the base64 dependency and use the HCL function. However, there were some encoding errors (quotation marks were not decoded) with it and I didn't look too deep since we support Linux and Mac were base64 is built-in and used with compatible flags (`-d`)

**Comment by msanft** [comment]
Regarding release pipeline changes, @malt3 should review that.

File: terraform/constellation-cluster/main.tf
Comment (elchead) [reply terraform/constellation-cluster/main.tf:84]: I thought it was nice to seperate conf and state configuration

**Comment by elchead** [comment]
Rebased and incorporated feedback. Would you test it also locally on Mac to see if it works as expected for you @msanft?

**Comment by msanft** [comment]
> Rebased and incorporated feedback. Would you test it also locally on Mac to see if it works as expected for you @msanft?

I don't have a MacOS machine set up for development right now unfortunately

File: .github/workflows/e2e-test-tf-module.yml
Comment (msanft) [inline .github/workflows/e2e-test-tf-module.yml:123]: I think we should rather add a default value in the `inputs` section, it doesn't need duplication there and is more accessible.

**Review by msanft — COMMENTED** [review]
TF specific changes LGTM except the comment. Didn't look at release pipeline or Go upgrade. Just ping me once ready for final review

**Review by malt3 — APPROVED** [review]
Only reviewed the changes made to the pipeline. They look good to me.
Apologies for the delay in reviewing from my side.

**Comment by msanft** [comment]
One thing that is broken here, but that will be addressed in #2566:

We cannot default to a `latest` version when downloading the CLI. This will break for users that have an outdated Terraform module, say of `v2.12.0`, and try to apply it at a point where `v2.13.0` is already released. `latest` will then point to  `v2.13.0`. If the  `v2.13.0`  CLI is incompatible with  the  `v2.12.0` module, this won't work.

We should rather let the user set the CLI version through a variable in the `constellation-cluster` module and set it during packaging of the module. I will take care of that in #2566.

File: .github/workflows/e2e-test-tf-module.yml
Comment (elchead) [reply .github/workflows/e2e-test-tf-module.yml:123]: Leo wanted exactly this since the default is provider specific and we don't want a region input for each csp.

**Comment by elchead** [comment]
> One thing that is broken here, but that will be addressed in #2566:
> 
> We cannot default to a `latest` version when downloading the CLI. This will break for users that have an outdated Terraform module, say of `v2.12.0`, and try to apply it at a point where `v2.13.0` is already released. `latest` will then point to `v2.13.0`. If the `v2.13.0` CLI is incompatible with the `v2.12.0` module, this won't work.
> 
> We should rather let the user set the CLI version through a variable in the `constellation-cluster` module and set it during packaging of the module. I will take care of that in #2566.

This is already fixed, we decided to inject the version during the release process:
https://github.com/edgelesssys/constellation/blob/cea6204b37c161169bc9675501305b6f2bbb68a8/dev-docs/workflows/release.md#L34-L35

**Review by msanft — APPROVED** [review]

**Pull request merged**
